### PR TITLE
Re-organize HSM validation to be more consistent with less duplication

### PR DESCRIPTION
**Pull request by rcritten**
hsm_validator() was more or less bolted in place late in the
development cycle in in order to catch some of the more common
problems: bad token name, bad password, etc.

There was a fair bit of duplication and had the side-effect of not
reading in the token password from the --token-password-file option
in some cases.

This patch also re-adds a lost feature where an exception is raised if
both the --token-password and --token-password-file options are passed
in.

Fixes: https://pagure.io/freeipa/issue/9603

### CODE REVIEW

**Comment by mrizwan93** [comment]
Tests are skipped. Tried with `update_packages: True` as well but no luck.

**Comment by flo-renaud** [comment]
@mrizwan93:
 you need to use @pki/master repo in order to have a PKI version that supports HSM on f39. You can have a look at https://github.com/freeipa/freeipa/blob/master/ipatests/prci_definitions/nightly_latest_pki.yaml for test definitions.

The other option is to use the f40 vagrant image that is being tested right now in https://github.com/freeipa/freeipa/pull/7383.

File: ipatests/prci_definitions/temp_commit.yaml [hunk 1]
```diff
@@ -53,6 +53,7 @@ jobs:
     job:
       class: Build
       args:
+        copr: '@pki/master'
```
Comment (rcritten) [inline ipatests/prci_definitions/temp_commit.yaml:56]: We're seeing more than a few very strange failures. The pki/master copr repo is 11.6.0. I wonder if we should instead test against 11.5.0 which is what is in F40+ right now. My freeipa copr repo has a 11.5.0 build in it.

**Comment by rcritten** [comment]
I know why the Azure test is failing. Working on it.

**Comment by flo-renaud** [comment]
Hi @rcritten 

The installer prompts for the password even if -U (unattended) is provided:
```
# ipa-server-install --domain ipa.test --realm IPA.TEST -a Secret123 -p Secret123 -U --token-name myhsmtoken --token-library-path /usr/lib64/pkcs11/libsofthsm2.so

The log file for this installation can be found in /var/log/ipaserver-install.log
==============================================================================
This program will set up the IPA Server.
Version 4.13.0.dev

This includes:
  * Configure a stand-alone CA (dogtag) for certificate management
  * Configure the NTP client (chronyd)
  * Create and configure an instance of Directory Server
  * Create and configure a Kerberos Key Distribution Center (KDC)
  * Configure Apache (httpd)
  * Configure SID generation
  * Configure the KDC to enable PKINIT

HSM token 'myhsmtoken' password: 
```

**Comment by flo-renaud** [comment]
- I also noticed that the man page for ipa-kra-install does not mention the new options 

```
  --token-password=TOKEN_PASSWORD
                        The password to the PKCS#11 token.
  --token-password-file=TOKEN_PASSWORD_FILE
                        The full path containing the PKCS#11 token password.
```

- on the master, running KRA install produces the following error:
```
# ipa-kra-install -U -p Secret123
'Values' object has no attribute 'token_name'
The ipa-kra-install command failed. See /var/log/ipaserver-kra-install.log for more information
```
-

**Comment by rcritten** [comment]
> ```
> # ipa-kra-install -U -p Secret123
> 'Values' object has no attribute 'token_name'
> The ipa-kra-install command failed. See /var/log/ipaserver-kra-install.log for more information
> ```

Can you provide the traceback you got? I wasn't able to reproduce this.

**Comment by flo-renaud** [comment]
- The issue with prompting for password in unattended mode is now solved:

```
# ipa-server-install --domain hsm.test --realm HSM.TEST -a Secret123 -p Secret123 -U --token-library-path /usr/lib64/pkcs11/libsofthsm2.so --token-name flohsmtoken

The log file for this installation can be found in /var/log/ipaserver-install.log
==============================================================================
This program will set up the IPA Server.
Version 4.13.0.dev

This includes:
  * Configure a stand-alone CA (dogtag) for certificate management
  * Configure the NTP client (chronyd)
  * Create and configure an instance of Directory Server
  * Create and configure a Kerberos Key Distribution Center (KDC)
  * Configure Apache (httpd)
  * Configure SID generation
  * Configure the KDC to enable PKINIT

HSM token password required
The ipa-server-install command failed. See /var/log/ipaserver-install.log for more information
```
- the ipa-kra-install man page is now documenting the new options
- I still see the issue with ipa-kra-install. Reproducer: install IPA with embedded CA and HSM, call `ipa-kra-install -U -p Secret123`. When the first KRA is installed, the code from install_check also needs to read token name and library path from ca.lookup_hsm_configuration(api) (in ipaserver/install/kra.py, line 73 with the else clause).

**Review by flo-renaud — COMMENTED** [review]
Hi @rcritten 
please find one inline comment

**Comment by flo-renaud** [comment]
@rcritten 
thanks for the updated patch. The ipa-kra-install is now also working for non-HSM setups.

You can remove the temp commit and I will ack when PRCI completes.

**Comment by rcritten** [comment]
master:

* 23de845987c0776f77d8b8caeabf51f312bca5a6 Re-organize HSM validation to be more consistent/less duplication
* a11c843adcd5947ef124fc418bfb3e0ac0750ae4 ipatests: tests related to --token-password-file
* 8d5461bea785c43a14725a2fb8f5f705758c54ed Include token password options in ipa-kra-install man page

**Pull request closed without merging**